controller/deployments_blackbox_test.go

```diff
@@ -587,3 +595,152 @@ func createOSIOClientMock(t minimock.Tester, spaceName string) *testcontroller.O
 	}
 	return osioClientMock
 }
+
+type testWatchItem struct {
+	Name   string
+	Object testInnerItem
+}
+
+type testInnerItem struct {
+	Number int
+}
+
+func TestWatchEnvironmentEvents(t *testing.T) {
+	clientGetterMock := testcontroller.NewClientGetterMock(t)
+	svc, ctrl, err := createDeploymentsController()
+	require.NoError(t, err)
+	ctrl.ClientGetter = clientGetterMock
+
+	t.Run("ok", func(t *testing.T) {
+		testItems := []testWatchItem{
+			testWatchItem{
+				Name: "one",
+				Object: testInnerItem{
+					Number: 2,
+				},
+			},
+			testWatchItem{
+				Name: "two",
+				Object: testInnerItem{
+					Number: 3,
+				},
+			},
+		}
+
+		mockKeyFunc := func(obj interface{}) (string, error) {
+			if v, ok := obj.(testWatchItem); ok {
+				return v.Name, nil
+			}
+			return "default", nil
+		}
+
+		kubeClientMock := testk8s.NewKubeClientMock(t)
+		kubeClientMock.WatchEventsInNamespaceFunc = func(p string) (r *cache.FIFO, r1 chan struct{}) {
+			fmt.Println("Watch events function called")
+			store := cache.NewFIFO(mockKeyFunc)
+			for _, item := range testItems {
+				store.Add(item)
+			}
+
+			return store, make(chan struct{})
+		}
+		kubeClientMock.CloseFunc = func() {}
+		clientGetterMock.GetKubeClientFunc = func(p context.Context) (kubernetes.KubeClientInterface, error) {
+			return kubeClientMock, nil
+		}
+		osioClientMock := testcontroller.NewOSIOClientMock(t)
+		clientGetterMock.GetAndCheckOSIOClientFunc = func(p context.Context) (controller.OpenshiftIOClient, error) {
+			return osioClientMock, nil
+		}
+
+		conn := WatchEnvironmentEventsDeploymentsOK(t, context.Background(), svc, ctrl, space.SystemSpace)
+
+		var buf []byte
+		for _, item := range testItems {
+			// buffer 256 is an arbitrary choice that fits the test items
+			// Manually unmarshal ws frame. Second element contains length
+			// Object is marshaled as JSON and placed in index 2:length+2
+			buf = make([]byte, 256)
+			conn.Read(buf)
+			frameLength := int(buf[1])
+			startPos := 2
+			endPos := startPos + frameLength
```

Comment:
It's a shame we have to handle WebSocket frames manually, but I don't see a way around it unless we somehow return a websocket.Conn instead of a net.Conn from WatchEnvironmentEventsDeploymentsOK. Maybe add a link to the RFC section that details the frame encoding [1]? I found this helpful to understand what's going on here.

[1] https://tools.ietf.org/html/rfc6455#section-5.2